Develop

fmnjose opened the pull request:
Final pull request to main branch

fmnjose committed:
feat: add commands support and solve table creation error

Changed how the program is ran. It now supports a command interface to execute operations.
Only table and user creation are implemented yet.

Segregated the db object initialization and the server logic.

Added a catch block to the db.none promise which creates the table, and now the program
returns an error saying that the table was not created because it already exists.

fmnjose committed:
fix: fix bug that lets the db promise hanging

Solved by adding a process exit to end of the promise chain (after the catch, to also terminate the process in case of error)

fmnjose committed:
Merge pull request #1 from fmnjose/feature-challenge1

Feature challenge1

fmnjose committed:
feat: Add constraint on DB to refuse duplicate users

Added a UNIQUE constraint to the login value on the CREATE TABLE statement

fmnjose committed:
Merge pull request #2 from fmnjose/feature-no-duplicates

feat: Add constraint on DB to refuse duplicate users

fmnjose committed:
feat: Add new fields to the Database and API

Added the fields 'location' and 'followers' to the github_users table
Added the information relative to all the table fields in the GithubUsers object and INSERT

fmnjose committed:
feat: Add users listing by location

Added the command listbylocation <location> to list users by location.
Added a new controller to take care of formating data using ramda functional programming.
The code needs some tweaks to be more broad in terms of the commands and options available, as well as data processing

fmnjose committed:
Merge pull request #3 from fmnjose/feature-new-fields

feat: new fields and users listing by location

fmnjose committed:
fix: Fix error with the db object

There was an error relative to the way the db object was being imported in the server.js file

fmnjose committed:
feat: Add new table with liked languages

Added a command to create a new table liked_languages which is a one-to-many relation between user_id's and languages
Added a command to add a liked language to a user

fmnjose committed:
feat: Add listing of all users by location and language

Added 3 forms of listing users:
	- list all
	- list all filtering by location
	- list all filtering by location AND language

It is not possible to filter only by language

Changed the data processing to output all the languages liked by a user

fmnjose committed:
Merge pull request #4 from fmnjose/feature-liked-languages

Feature liked languages

fmnjose committed:
doc: add documentation

Added comments to all functions

Started working on README

fmnjose committed:
doc: Finish README and documentation

Ready to be merged to main branch

fmnjose committed:
doc: Fix README

fmnjose committed:
doc: Fixed last question answer

jcristovao commented on src/command-dispatcher.ts:
```diff
@@ -0,0 +1,82 @@
+/* Entry point of the program.
+   Processes the commands received and calls
+   server functions
+ */
```
Hey @fmnjose , can you modify the code so that after some operations, like `createUser` or `addLikedLanguage` a call is made to get all the entries on the database, and show them?
Besides the code that implements each of those functionalities, what did you have to change in order to guarantee the proper sequence of operations?

fmnjose replied:
I'm trying to solve this issue together with https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r840358555 ,
but I'm struggling with the bluebird library.

My idea was to have all the functions from server.ts as promises, and dispatcher.js would call them and terminate the program when the async call is finished ( either resolved or rejected).

This would also allow to chain functions from server.ts. For example, when executing the createUser function, i would chain to it the getUsers function, obtaining a promise chain that will create the user and only list all entries after the user creation (considering it is successful).

I will keep chasing this, but will commit the changes to the other comments in the meantime.

fmnjose replied:
Bluebird promises finally clicked!
Will commit changes done to support async functions and db calls chaining.

jcristovao commented on src/command-dispatcher.ts:
```diff
@@ -0,0 +1,82 @@
+import {  } from "./server";
```
Why do you have an empty import first, and then a `require`? 
Would it be possible to use only `import` instead of `require`?

jcristovao commented on src/server.ts:
```diff
@@ -0,0 +1,156 @@
+/**Creates the table with name LIKED_LANGUAGES_NAME to store
+ * the relationship between users and languages they like
+ */
+const createLikedLanguagesTable = () => {
```
There is a simpler syntax for adding the reference on `user_id`...
Also, we really prefer that these long lines are split into several lines, to increase readability

fmnjose replied:
The next commit will have a simpler syntax for the foreign key definition.
The long lines will also be split

jcristovao commented on src/server.ts:
```diff
@@ -0,0 +1,156 @@
+          CONSTRAINT fk_user FOREIGN KEY(user_id) REFERENCES ${USERS_TABLE_NAME}(id))`)
+    .catch((e: Error) => {
+      errorHandler(e);
+    })
```
Only the main program should issue a `process.exit`, the remaining functions should just return something

jcristovao commented on src/server.ts:
```diff
@@ -0,0 +1,156 @@
+ * @param language language name
+ */
+const addLikedLanguage = (user: string, language: string) => {
+  db.one(
```
What happens if the user provides a malicious user string?

fmnjose replied:
The user could atack the database with using sql injection.
I have changed the code to use preprared statements from the pg-promise library on every query that depends on user input.
Will commit the changes together with the other ones

fmnjose committed:
fix: Change database querying

Changed the database calls to use pg-promise prepared statements

Changed the liked_languages creation query to simplify the foreign key definition

Changed the getAllUsers and getUsersByLocation queries to use Left Join, in order to list all users even if they don't have any liked language defined

The output string has been tweaked to include the user Location

fmnjose reviewed (commented):
@jcristovao this commit solves issues:

- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r840355775
- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r840357879
- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r840359215

fmnjose committed:
feat: Convert functions to async calls

Changed the server functions to return promises. These are called by the dispatcher, which catches errors and terminates the process.

Chaining operations is now possible. When adding a new user or liked language, the program outputs the list of users already updated with the newdata

fmnjose committed:
fix: Fixed code that allowed a null user_id in the liked_languages table

fmnjose reviewed (commented):
@jcristovao this commit addresses the following issues:

- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r840354997
- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r840358555

dmsapereira commented on src/command-dispatcher.ts:
```diff
@@ -0,0 +1,82 @@
+import {  } from "./server";
+
+const {createUsersTable, createUser, createLikedLanguagesTable, 
```
These imports become much more readable (and easy on the eyes) if you do something like:
```
const {
  a,
  b,
  c,
  d
} = require('./whatever');
```
It also makes the 80 character limit issue no longer a thing (unless you name functions like in Java, which you shouldn't)

fmnjose replied:
I can't seem to find the empty import line in the develop branch code. I don't understand why is it appearing for you.

Ok, understood :)
I will change that

dmsapereira commented on src/command-dispatcher.ts:
```diff
@@ -0,0 +1,82 @@
+   Processes the commands received and calls
+   server functions
+ */
+function dispatchCommand(command: string[]){
```
Just how you declare a parameter type, you can declare a function return type to make it extra explicit, and to make sure typescript isn't assuming some dumb types out of this.
Something like: `function foo(bar: string) : Promise<any>{...}`
As a bonus TS/ESLint will warn you if you're doing something wrong, so it's like a soft static typing!

fmnjose replied:
That makes sense. Will change it as well on the next commit

dmsapereira commented on src/command-dispatcher.ts:
```diff
@@ -0,0 +1,82 @@
+        "# " + commandsList.ADD_LIKED_LANGUAGE + " <user_name> <language> -> Add <language> to the list of <user_name>'s liked languages\n\n",
+        "# " + commandsList.LIST_USERS + " [optional: location] [optional: language] -> List all users in DB. If location is passed, filters by location. If location and language are passed, parses by both filters. Never parses only by language\n"
+    )
+}
```
You can use backticks for multi-line strings and you even get string formatting/replacement!
This removes the issue of the 80 character limit here and makes it easier on the eyes for stuff like:
```
console.log("What's your name?\n My name is " + name + " !");
```
VS
```
console.log(
`
What's your name?
My name is ${name}!
`
);
```

More on it [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)!

dmsapereira commented on src/data-processor.ts:
```diff
@@ -0,0 +1,45 @@
+    let outputArray = R.filter((line) => line != "",outputFormat(data));
+  
+    R.map((out) => console.log("##",out))(outputArray);
+}
```
To increase this function's readability you could declare the `map` function **outside** the function and just call it inside of it.
Something like:
```
const mapFn = val => console.log(val);
.
.
.

let outputFormat = R.map(mapFn);
```
It'll become much more readable, and less like something close to a [callback hell staircase](http://callbackhell.com/)!

dmsapereira commented on src/data-processor.ts:
```diff
@@ -0,0 +1,45 @@
+        };
+    
+    R.map((user) => {getLanguageOnMatch(user)},usersList);
+
```
Considering its behaviour this could be a [forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach), instead of a `Ramda.map`.
Native functions are always better than imported ones!

fmnjose replied:
Thank you for the notes. 
I will change this and https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r841628886

dmsapereira reviewed (changes requested):
Please address mine and João's comments.
There are some minor nitpicks, but there are some major flaws like the SQL formatting and Promise returns.

fmnjose committed:
fix: Address issues with code readibility

Changed the commandList enum to class. The class has a constructor with two arguments: the command input string and the instructions on how to use it. The class has all the commands as static instances of itself, and an override of the toString() method, to print the help menu entry.

The dispatchCommand function has now a return type of Promise<any> and the switche's default has been changed to also return a promise, as the program was throwing an exception when trying to execute the .then on a wrong command input.

Import statement for the server module functions on the dispatcher was also changed for readibility.

The data-processor module was changed to increase readibility.

fmnjose reviewed (commented):
@dmsapereira this commit addresses:

- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r841613108
- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r841617102
- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r841623825
- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r841628886
- https://github.com/fmnjose/lovelystay-test/pull/5#discussion_r841631273
